[Tools][GTK][WPE] MVT test suite: integrate run-mvt-tests with webkitpy facilities and support test expectations

clopez opened the pull request:
#### 566fb370f6b383f85b715686e67a39098bf8dc73
<pre>
[Tools][GTK][WPE] MVT test suite: integrate run-mvt-tests with webkitpy facilities and support test expectations
<a href="https://bugs.webkit.org/show_bug.cgi?id=284916">https://bugs.webkit.org/show_bug.cgi?id=284916</a>

Reviewed by Adrian Perez de Castro.

This patch improves the script run-mvt-tests with the following features:

 - Support running with WebKitGTK.
 - For WPE support running both with MiniBrowser and Cog.
 - Support running without Flatpak (system libs) as well as with JHBuild.
 - Automatic enabling of headless mode via command-line switch
 - Allow to run the whole MVT tests automatically (all test suites).
   This is the default now.
 - Support for test expectations in order to mark expected failures, including
   support for auto-generating new expectations based on the test run results.
 - Improved logging (with options to enable debug messages or to run quiet only
   logging unexpected failures).
 - Better integration with webkitpy facilities (by using the new run-webdriver
   script that integrates with the webkitpy port driver to set the environment
   for the webdriver and browser rather than manually having to set environment
   variables inside the script).
 - Improved handling of corner-cases where the WebDriver hangs.
 - Automatic cleaning of leftover process after the test run.
 - Automatic randomization of the WebDriver port used for tests, making easier
   to trigger several test runs on the same host at the same time.

On top of that, this also fixes the following:

 - run-minibrowser: was not correctly setting the exit code on failure.
 - create a new run-webdriver helper script to start the webdriver from
   a given platform/port built at WebKitBuild.
 - autoinstalled selenium: upgrade to a newer version of selenium and properly
   install the new websocket-client subdependency with the auto-install logic.
   Note: This new version of Selenium is required to create the driver with a
   custom path and arguments for the webdriver.

* Tools/Scripts/libraries/resultsdbpy/resultsdbpy/__init__.py:
* Tools/Scripts/run-minibrowser:
* Tools/Scripts/run-mvt-tests:
(parse_args):
(fatal): Deleted.
(options): Deleted.
(suite_has_started): Deleted.
(suite_has_finished): Deleted.
(analyze_results): Deleted.
(print_tests_failed): Deleted.
(WPEWebDriver): Deleted.
(WPEWebDriver.__init__): Deleted.
(WPEWebDriver.start): Deleted.
(WPEWebDriver.stop): Deleted.
(WPEWebDriver._isRunning): Deleted.
(WPEWebDriver._isFlatpakSDK): Deleted.
(WPEWebDriver._isJHBuild): Deleted.
(WPEWebDriver._run): Deleted.
(WPEWebDriver._run.dict_flat): Deleted.
(main): Deleted.
* Tools/Scripts/run-webdriver: Added.
(system_has_ipv6):
(main):
* Tools/Scripts/webkitpy/__init__.py:
* Tools/Scripts/webkitpy/port/base.py:
(Port.run_webdriver):
* Tools/Scripts/webkitpy/port/glib.py:
(GLibPort._get_crash_log):
(GLibPort):
(GLibPort.setup_environ_for_webdriver):
(GLibPort.run_webdriver):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort):
(WPEPort.setup_environ_for_webdriver):
* Tools/Scripts/webkitpy/webdriver_tests/webdriver_driver_wpe.py:
(WebDriverWPE.browser_env):
* Tools/Scripts/webkitpy/webdriver_tests/webdriver_test_runner_selenium.py:
* Tools/gtk/MVT_TestExpectations_Release.json: Added.
* Tools/wpe/MVT_TestExpectations_Release.json: Added.

Canonical link: <a href="https://commits.webkit.org/288842@main">https://commits.webkit.org/288842@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c18f3465202348f9517f7b5126ba851806ceaea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4362 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39025 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89776 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35688 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86722 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4451 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12335 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/65877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23705 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87682 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/3380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/76921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/84045 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/31133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34763 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/31934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91151 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11976 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/84272 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12203 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/72725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/73475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17844 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/16285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/3420 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13174 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11928 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11762 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15256 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13508 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -30,317 +30,350 @@ if os.path.isdir(os.path.join(scripts_dir, 'webkitpy')):
+    platform = parser.add_mutually_exclusive_group(required=True)
+    platform.add_argument('--gtk', action='store_const', dest='platform', const='gtk', help='Use the GTK port')
+    platform.add_argument('--wpe', action='store_const', dest='platform', const='wpe', help='Use the WPE port')
+    parser.add_argument("--extra-browser-args", help="Extra arguments to pass to the browser (run-minibrowser)")
```
This may easily run into quoting awkwardness. How about handling `--` and taking the leftover command line arguments after that as the ones to be passed down to `run-minibrowser`? Many command line tools work that way.

aperezdc commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -30,317 +30,350 @@ if os.path.isdir(os.path.join(scripts_dir, 'webkitpy')):
+        if 'timeout' not in kwargs:
+            kwargs['timeout'] = timeout
+        return original_request(self, method, url, *args, **kwargs)
+    original_request = urllib3.request.RequestMethods.request
```
I'm somewhat surprised that it's not needed to put this line before the `patched_request` is defined. Could you move it up to better indicate the intention even if it currently works?

clopez replied:
AFAIK Python functions can access variables declared above its own scope, it doesn't matter in the line when you declare the variable, what matters is if the variable is already defined when you call the function.
But I can declare it above the function declaration if that makes the code more clear.

aperezdc commented on Tools/Scripts/run-webdriver:
```diff
@@ -0,0 +1,74 @@
+        raise
+
+def main(argv):
+
```
Remove this empty line.

```suggestion
```

aperezdc reviewed (commented):
Nice improvements overall, I left a couple of comments 👌🏼

dpino reviewed (approved):
I reviewed the code. I don't have any suggestions, it all looks to me. I also pulled the PR and tested it. It worked as expected.

lauromoura commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -30,317 +30,355 @@ if os.path.isdir(os.path.join(scripts_dir, 'webkitpy')):
+        for run_minibrowser_arg in run_minibrowser_args:
+            options.add_argument(run_minibrowser_arg)
+        if browser_name:
+            options.capabilities['browserName'] = browser_name
```
```suggestion
            options.set_capability("browserName", browser_name)
```

clopez replied:
ok, nice. I will apply it

lauromoura reviewed (commented):
LGTM. Just a nit about setting the capabilities, as Selenium added some helpers to hide the capability dictionary.

aperezdc commented on Tools/Scripts/run-mvt-tests:
```diff
@@ -30,317 +30,355 @@ if os.path.isdir(os.path.join(scripts_dir, 'webkitpy')):
+    if num_suites > 1:
+        _log.info('######################')
+        _log.info('# TOTAL TEST SUMMARY #')
+        _log.info('######################')
```
“Total test summary” sounds odd in English. I think you want to just write “Summary” here 😉 

```suggestion
        _log.info('###########')
        _log.info('# SUMMARY #')
        _log.info('###########')
```

aperezdc reviewed (approved):
LGTM, with a nit you may want to take into account before landing 👌🏼

clopez committed:
[Tools][GTK][WPE] MVT test suite: integrate run-mvt-tests with webkitpy facilities and support test expectations
https://bugs.webkit.org/show_bug.cgi?id=284916

Reviewed by Adrian Perez de Castro.

This patch improves the script run-mvt-tests with the following features:

 - Support running with WebKitGTK.
 - For WPE support running both with MiniBrowser and Cog.
 - Support running without Flatpak (system libs) as well as with JHBuild.
 - Automatic enabling of headless mode via command-line switch
 - Allow to run the whole MVT tests automatically (all test suites).
   This is the default now.
 - Support for test expectations in order to mark expected failures, including
   support for auto-generating new expectations based on the test run results.
 - Improved logging (with options to enable debug messages or to run quiet only
   logging unexpected failures).
 - Better integration with webkitpy facilities (by using the new run-webdriver
   script that integrates with the webkitpy port driver to set the environment
   for the webdriver and browser rather than manually having to set environment
   variables inside the script).
 - Improved handling of corner-cases where the WebDriver hangs.
 - Automatic cleaning of leftover process after the test run.
 - Automatic randomization of the WebDriver port used for tests, making easier
   to trigger several test runs on the same host at the same time.

On top of that, this also fixes the following:

 - run-minibrowser: was not correctly setting the exit code on failure.
 - create a new run-webdriver helper script to start the webdriver from
   a given platform/port built at WebKitBuild.
 - autoinstalled selenium: upgrade to a newer version of selenium and properly
   install the new websocket-client subdependency with the auto-install logic.
   Note: This new version of Selenium is required to create the driver with a
   custom path and arguments for the webdriver.

* Tools/Scripts/libraries/resultsdbpy/resultsdbpy/__init__.py:
* Tools/Scripts/run-minibrowser:
* Tools/Scripts/run-mvt-tests:
(parse_args):
(fatal): Deleted.
(options): Deleted.
(suite_has_started): Deleted.
(suite_has_finished): Deleted.
(analyze_results): Deleted.
(print_tests_failed): Deleted.
(WPEWebDriver): Deleted.
(WPEWebDriver.__init__): Deleted.
(WPEWebDriver.start): Deleted.
(WPEWebDriver.stop): Deleted.
(WPEWebDriver._isRunning): Deleted.
(WPEWebDriver._isFlatpakSDK): Deleted.
(WPEWebDriver._isJHBuild): Deleted.
(WPEWebDriver._run): Deleted.
(WPEWebDriver._run.dict_flat): Deleted.
(main): Deleted.
* Tools/Scripts/run-webdriver: Added.
(system_has_ipv6):
(main):
* Tools/Scripts/webkitpy/__init__.py:
* Tools/Scripts/webkitpy/port/base.py:
(Port.run_webdriver):
* Tools/Scripts/webkitpy/port/glib.py:
(GLibPort._get_crash_log):
(GLibPort):
(GLibPort.setup_environ_for_webdriver):
(GLibPort.run_webdriver):
* Tools/Scripts/webkitpy/port/gtk.py:
(GtkPort):
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort):
(WPEPort.setup_environ_for_webdriver):
* Tools/Scripts/webkitpy/webdriver_tests/webdriver_driver_wpe.py:
(WebDriverWPE.browser_env):
* Tools/Scripts/webkitpy/webdriver_tests/webdriver_test_runner_selenium.py:
* Tools/gtk/MVT_TestExpectations_Release.json: Added.
* Tools/wpe/MVT_TestExpectations_Release.json: Added.

Canonical link: https://commits.webkit.org/288842@main
